src/highdicom/legacy/sop.py

```diff
@@ -25,384 +31,506 @@
-
-    # Shared Functional Groups
-    shared_item = Dataset()
-
-    # Pixel Measures (M)
-    pixel_measures_item = Dataset()
-    pixel_measures_item.PixelSpacing = ref_ds.PixelSpacing
-    pixel_measures_item.SliceThickness = ref_ds.SliceThickness
-    try:
-        pixel_measures_item.SpacingBetweenSlices = \
-            ref_ds.SpacingBetweenSlices
-    except AttributeError:
-        pass
-    shared_item.PixelMeasuresSequence = [
-        pixel_measures_item,
-    ]
-
-    # Plane Orientation (Patient) (M)
-    plane_orientation_item = Dataset()
-    plane_orientation_item.ImageOrientationPatient = \
-        ref_ds.ImageOrientationPatient
-    shared_item.PlaneOrientationSequence = [
-        plane_orientation_item,
-    ]
-
-    shared_item.UnassignedSharedConvertedAttributesSequence = [
-        unassigned_shared_ca_item,
-    ]
-    mf_dataset.SharedFunctionalGroupsSequence = [
-        shared_item,
-    ]
-
-    for i, ca_item in enumerate(unassigned_perframe_ca_items):
-        perframe_items[i].UnassignedPerFrameConvertedAttributesSequence = [
-            ca_item,
-        ]
-    mf_dataset.PerFrameFunctionalGroupsSequence = perframe_items
+            n1: ndarray = slice1.get_normal_vector()
+            n2: ndarray = slice2.get_normal_vector()
+            for el1, el2 in zip(n1, n2):
+                if abs(el1 - el2) > tolerance:
+                    return False
+            return True
+
+
+class ـPerframeFunctionalGroup(DataElementSequence):
 
-    mf_dataset.AcquisitionContextSequence = []
+    """A sequence class for perframe functional group"""
 
-    # TODO: Encapsulated Pixel Data with compressed frame items.
+    def __init__(self, number_of_frames: int) -> None:
+        """
+        Parameters
+        ----------
+        number_of_frames: int
+            The perframe functional group sequence will have items equal to
+            the whole number of frames
+        """
+        super().__init__()
+        for i in range(0, number_of_frames):
+            item = Dataset()
+            self.append(item)
 
-    # Create the Pixel Data element of the mulit-frame image instance using
-    # native encoding (simply concatenating pixels of individual frames)
-    # Sometimes there may be numpy types such as ">i2". The (* 1) hack
-    # ensures that pixel values have the correct integer type.
-    mf_dataset.PixelData = b''.join([
-        (ds.pixel_array * 1).data for ds in sf_datasets
-    ])
 
-    return mf_dataset
+class ـSharedFunctionalGroup(DataElementSequence):
 
+    """A sequence class for shared functional group"""
 
-class LegacyConvertedEnhancedMRImage(SOPClass):
+    def __init__(self) -> None:
+        """Construncts a shared frame functional group holding only one item
 
-    """SOP class for Legacy Converted Enhanced MR Image instances."""
+        """
+        super().__init__()
+        item = Dataset()
+        self.append(item)
+
+
+class FrameSet:
```

Comment:
Please make this class "private", i.e., rename to `_FrameSet`

Reply:
I'm afraid we can't make FrameSet and FrameSetCollection private. We use both classes to detect potential multi-frame sets within every candidate collection before feeding them into LegacyConvertedEnhanced images.

Reply:
The overall pattern throughout the library is that constructors of classes derived from `SOPClass` accept the source images as `Sequence[pydicom.dataset.Dataset`. That's also how the `LegacyConvertedEnhanced*Image` classes were implemented. It looks like you updated the code to remain backwards compatible with the public API. It should now be possible to make these classes private, shouldn't it?

Reply:
The common use case to set apart all frames that belong to a potential MF dicom instance before feeding the list into SOPClass to avoid any failure. If we don't do that the conversion will not proceed and will crash. Of course, it is not necessary to do this when it is guaranteed that the input single-frame list belongs to one MF instance, in our case though, it is not.

Reply:
Well, sure that it's fine for the conversion to crash and it does. But, what should users do if they want to convert a mixed series? Don't they need a separate tool to figure out which single frames are potential multi-frame?? If they develop this tool on their own they need to duplicate the logic of the Frameset class.